### ISSUE TITLE
MySQL 8 support for Drush 8

### ISSUE THREAD

**Issue by Ayesh** [drush-ops/drush#3830]
**Describe the bug**
MySQL 8 was released a few months back, Drush quits with fatal errors when tried to run with MySQL due to various reasons. 

I suggest that we follow the D8 and D7 issued and upgrade `dbtng` component accordingly. 

 - https://www.drupal.org/project/drupal/issues/2986452
 - https://www.drupal.org/project/drupal/issues/2978575

### System Configuration
| Q               | A
| --------------- | ---
| Drush version?  | 8.x
| Drupal version? | 7.61
| PHP version     | 7.3
| MySQL version     | 8.0
| OS?             | Ubuntu 18.10

**Additional information**

**Comment by weitzman** [comment]
So does Drush need changes or just Drupal? I expect that Drush9 is the focus for this issue as Drush8 is in maint support only.

**Comment by weitzman** [comment]
What are the fatal errors? We do want to be compatible with new Mysql versions, of course.

**Comment by MaksOta** [comment]
These are the errors that i am getting when trying to do anything with drush on mysql 8:

`WD php: PDOException: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to    [error]
your MySQL server version for the right syntax to use near 'system
WHERE  (type='module' AND status='1')' at line 3: SELECT system.name AS name
FROM
{system} system
WHERE  (type=:type AND status=:status) ; Array
(
    [:type] => module
    [:status] => 1
)
 in drush_db_select() (line 99 of /Users/mask/.composer/vendor/drush/drush/includes/dbtng.inc).
PDOException: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near &#039;system
WHERE  (type=&#039;module&#039; AND status=&#039;1&#039;)&#039; at line 3: SELECT system.name AS name
FROM
{system} system
WHERE  (type=:type AND status=:status) ; Array
(
    [:type] => module
    [:status] => 1
)
 in drush_db_select() (line 99 of /Users/mask/.composer/vendor/drush/drush/includes/dbtng.inc).
Drush command terminated abnormally due to an unrecoverable error.`

**Comment by weitzman** [comment]
Please test with drush9. Drush8 is maint only.

**Comment by MaksOta** [comment]
I see

But drush9 doesn't work with Drupal 7. And with mysql 8 i am only having drush problems on d7. 

So i guess this is a "wont fix"?

**Comment by weitzman** [comment]
The title of the issue claims that the bug affects drush9 but the example is from 8. 

Drush8 could possibly get fixed if someone makes a PR.

**Comment by verbeeksteven** [comment]
You can get Drush 8 working with Drupal 7 if you apply the patch at  https://www.drupal.org/project/drupal/issues/2978575 and follow comment https://www.drupal.org/project/drupal/issues/2978575#comment-12731119 (add it to line 365 in includes/database/mysql/database.inc)

**Comment by vizdy** [comment]
weitzman   Fix to support mysql8.x with drupal 7 is needed in drush 8.x.  Though now drupal 7 works with mysql8.x with certain patches, drush does not work because drush8.x does not support mysql8.

Can you please see if someone can fix it?
==================================================================
WD php: PDOException: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds[error]
to your MySQL server version for the right syntax to use near 'system
WHERE  (type='module' AND status='1')' at line 3: SELECT system.name AS name
FROM 
{system} system
WHERE  (type=:type AND status=:status) ; Array
(
    [:type] => module
    [:status] => 1
)
 in drush_db_select() (line 99 of /home/fluser/.composer/vendor/drush/drush/includes/dbtng.inc).
PDOException: SQLSTATE[42000]: Syntax error or access violation: 1064 You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near &#039;system
WHERE  (type=&#039;module&#039; AND status=&#039;1&#039;)&#039; at line 3: SELECT system.name AS name
FROM 
{system} system
WHERE  (type=:type AND status=:status) ; Array
(
    [:type] => module
    [:status] => 1
)
 in drush_db_select() (line 99 of /home/fluser/.composer/vendor/drush/drush/includes/dbtng.inc).
Drush command terminated abnormally due to an unrecoverable error.

**Comment by weitzman** [comment]
I'm not sure how to "see" if someone can fix it. The whole world is welcome to fix it, and post a PR in this project. There are no paid people assigned to this project.

**Comment by amardeokar23** [comment]
Apply patch to drupal 7.x mentioned in https://github.com/drush-ops/drush/issues/3830#issuecomment-467049608 to support mysql 8.

To fix issue in drush we need to change code from function drush_db_select() which intern calls db_select() and uses table name table alias. ( https://github.com/drush-ops/drush/blob/8.x/includes/dbtng.inc )
Patch applied on drupal only escapes table name specified within "{}" . It doesn't escape reserved keyword used in select fields or table alias.
To fix above issue below is the patch.

```
--- a/includes/dbtng.inc        2019-04-18 12:56:57.067937807 +0200
+++ b/includes/dbtng.inc        2019-04-18 12:57:13.991909890 +0200
@@ -85,8 +85,8 @@
         $fields = array($fields);
       }
     }
-    $query = db_select($table, $table)
-      ->fields($table, $fields);
+    $query = db_select($table, 't')
+      ->fields('t', $fields);
     if (!empty($where)) {
       $query = $query->where($where, $args);
     }

```

**Comment by greg-1-anderson** [comment]
I just installed the recommended MySQL on MacOS 10.14, so I ended up with MySQL 8 without really intending it.  All of the Drush 9 tests pass with MySQL 8, so I think we're good there.

Under Drush 8, though, the tests freeze trying to set up the test Drupal site. I can run `site-install` from the command line, though, at least for Drupal 8. I cannot successfully run `drush site-install` on an unpatched Drupal 7, though. I have not tried any of the patches yet.

**Comment by mustanggb** [comment]
Just ran into this, and I agree with @amardeokar23 that this is the correct/best fix.

I've created a pull request for Drush 8 at #4313.

**Comment by joejoseph00** [comment]
I just updated the Pull Request with background info.   https://github.com/drush-ops/drush/pull/4313

**Comment by Ayesh** [comment]
Fixed in #4313. Thank you 🙏.
